tests/dags/test_invalid_dup_dependency.py

```diff
@@ -0,0 +1,34 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+from datetime import datetime
+
+from airflow import DAG
+from airflow.operators.empty import EmptyOperator
+
+with DAG(
+    "test_invalid_dup_dependency",
+    start_date=datetime(2021, 1, 1),
+    schedule="@once",
+    unique_dependencies=True,
+):
+    task_1 = EmptyOperator(task_id="one")
+    task_2 = EmptyOperator(task_id="two")
+
+    task_1 >> task_2
+    task_1 >> task_2
```

Comment:
Not sure I understand what is the problem with that case?
It's like someone write:
```
X = 4
X = 4
```

If the issue is visability why not register the duplicated dependencies in the warning?
This just doesn't feel like something worth a new parameter. The only time this larameter will take effect is when you deploy new version of code and I want to believe that when users develop new code they check the logs for thier recent change thus making th warning more informative sounds like the better way to go.

WDYT?

Reply:
As in the discussion I linked this is at probably best a code smell and at worst a logical error. It can also be very hard to debug when your DAG is 1000s of lines long and/or your tasks are generated by utility functions.

In your example:

```python
X = 1
X = 1
```

Likely the author either doesn't understand assignment in Python or they made a logical error likely meaning something like:

```python
X = 1
Y = 1
```

And if you made this mistake you're likely to get some runtime error, probably an Unbound Variable error.

But in this case when authoring an Airflow DAG no exception is thrown and you likely won't see the log message because DAG scripts themselves don't often setup a logger.

Further even if the DAG author see's the log warning there's no indication of where in the DAG construction it appeared from. Whereas the exception will give you exactly where the task duplicate task was added in your code.

I agree that it would be better to expose this without having to add an extra argument, happy to take suggestions and rework this idea.

Reply:
I understand you position but to me this is something that should be caught by linter not by Airflow core.

I think this is more in the area of tools like
https://github.com/feluelle/airflint

Accepting this one may also open the window to more requests of similar issues and I don't know if we should. I'm not saying I'm against just that it requires some discussion. But hey I'm 1 voice lets see what others think.

Reply:
cc @feluelle as the autor of linter tool would be nice to hear your perspective

Reply:
I think that's a reasonable viewpoint, but it has 2 practical problems:

1. As far as I'm aware a mature Airflow linter is a hypothetical not a reality
2. A linter would not be able to catch dynamically adding dependencies (e.g. from config/data files)

I would be happy to try and expose this in some either way, maybe via a test command or method?

Reply:
I think it would be useful for the development of airflow and debugging of airflow DAGs to log the registration of dependencies.

Reply:
Yes. Warning on dependencies would be more then fine I think. If you build your generator and would like to make sure that there are no duplicates in the tool generation you could catch the warning in the test code for the tool. Seems a bit excessive to raise an exception on that.

Reply:
Cool so we agree on informative warning.
@notatallshaw will you please apply? It might get to 2.5.3 if we will be fast enough

Reply:
I'm confused, there is an existing `self.log.warning` as you can see in the PRs diff and I also mentioned in this discussion. Do you mean turn it into a `warnings.warn` so that users can use flags to escalate it to an error if they chose?

Unfortunately I am having some confusing behavior on my side which is in my work environment the logger warning usually doesn't emit, but I don't see that in my home environment, so I think I'm going to have to close this PR until I understand that.

Still I believe though that the log warning isn't very helpful debugging the issue because it can't highlight where in your DAG generation this happened.